roman_kuprii/graph_generator.cpp

```diff
@@ -120,50 +136,115 @@ void add_yellow_edges(Graph& work_graph) {
 }
 
 void paint_edges(Graph& work_graph) {
-  add_blue_edges(work_graph);
-  add_green_edges(work_graph);
-  add_red_edges(work_graph);
-  add_yellow_edges(work_graph);
+  std::mutex add_edges_mutex;
+  std::thread blue_thread([&work_graph, &add_edges_mutex]() {
+    add_blue_edges(work_graph, add_edges_mutex);
+  });
+  std::thread green_thread([&work_graph, &add_edges_mutex]() {
+    add_green_edges(work_graph, add_edges_mutex);
+  });
+  std::thread red_thread([&work_graph, &add_edges_mutex]() {
+    add_red_edges(work_graph, add_edges_mutex);
+  });
+  std::thread yellow_thread([&work_graph, &add_edges_mutex]() {
+    add_yellow_edges(work_graph, add_edges_mutex);
+  });
+  blue_thread.join();
+  green_thread.join();
+  red_thread.join();
+  yellow_thread.join();
 }
 
-void generate_new_vertices(Graph& work_graph,
-                           uni_cpp_practice::graph_generation::Params params) {
+void generate_gray_branch(Graph& work_graph,
+                          uni_cpp_practice::GraphGenerator::Params& params,
+                          std::mutex& modify_mutex,
+                          const VertexId& parent_vertex_id,
+                          int current_depth) {
   int depth = params.depth;
   int new_vertices_num = params.new_vertices_num;
-  for (int current_depth = 0; current_depth <= depth; current_depth++) {
+  VertexId new_vertex_id;
+  {
+    const std::lock_guard lock(modify_mutex);
+    work_graph.add_vertex();
+    new_vertex_id = work_graph.get_vertices().back().get_id();
+    work_graph.connect_vertices(parent_vertex_id, new_vertex_id, true);
+  }
```

Comment:
1) лучше сразу инициализировать в константу
2) что за жесть `new_vertex_id = work_graph.get_vertices().back().get_id();`
```suggestion
  const VertexId new_vertex_id = []() {
    const std::lock_guard lock(modify_mutex);
    const auto new_vertex_id = work_graph.add_vertex();
    work_graph.connect_vertices(parent_vertex_id, new_vertex_id, true);
    return new_vertex_id;
  }()
```